week-03/assessment/solar-farm/solar-farm/src/main/java/learn/solar/data/PanelFileRepository.java

```diff
@@ -0,0 +1,160 @@
+        // create a list of log entries
+        List<Panel> result = new ArrayList<>();
+        // try with resources our buffered reader
+        try(BufferedReader reader = new BufferedReader(new FileReader(filePath))){
+            // skip the first line (header line)
+            reader.readLine();
+
+            // loop through our file and read each entry
+            for(String line = reader.readLine(); line != null; line = reader.readLine()){
+                Panel panel = deserialize(line);
+                result.add(panel);
+            }
+        }catch(FileNotFoundException ex){
+            // leave blank
+        }catch (IOException ex){
+            throw new DataException("Could Not Open File Path: " + filePath, ex);
+        }
+        return result;
+    }
+
+    private Panel deserialize(String line) {
+        String[] panelList = line.split(",");
+        if (panelList.length < 7) {
+            return null;
+        }
+        int id = Integer.parseInt(panelList[0]);
+        String section = panelList[1];
+        int row = Integer.parseInt(panelList[2]);
+        int column = Integer.parseInt(panelList[3]);
+        int year = Integer.parseInt(panelList[4]);
+        Material material = Material.valueOf(panelList[5]);
+        boolean isTracking = panelList[6].equals("Yes");
+        Panel panel = new Panel(id, section, row, column, year, material, isTracking);
+        return panel;
+    }
+
+    //- `public List<Panel> findBySection(String)` -- ???finds all Panels in a section, uses the private `findAll` method
+
+    private int getNextId(List<Panel> panels) {
+        int maxId = 0;
+        for (Panel panel : panels) {
+            if (maxId < panel.getId()) {
+                maxId = panel.getId();
+            }
+        }
+        return maxId + 1;
+    }
+
+    private void writeToFile(List<Panel> panels) throws DataException {
+        try(PrintWriter writer = new PrintWriter(filePath)){
+            writer.println("id,section,row,column,yearinstalled,material,istracking");
+
+            for(Panel panel : panels){
+                String line = serialize(panel);
+                writer.println(line);
+            }
+        }catch(IOException ex){
+            throw new DataException("Could not write to file path" + filePath);
+        }
+    }
+    private String serialize(Panel panel){
+        StringBuilder buffer = new StringBuilder();
+        buffer.append(panel.getId()).append(DELIMITER);
+        buffer.append(panel.getSection()).append(DELIMITER);
+        buffer.append(panel.getRow()).append(DELIMITER);
+        buffer.append(panel.getColumn()).append(DELIMITER);
+        buffer.append(panel.getInstallationYear()).append(DELIMITER);
+        buffer.append(panel.getMaterial()).append(DELIMITER);
+        if(panel.isTracking()){
+            buffer.append("Yes");
+        }else {
+            buffer.append("No");
+        }
+        return buffer.toString();
+    }
```

Comment:
```suggestion
buffer.append(solarPanel.isTracking());
```
no need for an if statement here